ccl/serverccl: TestResetIndexUsageStatsRPCForTenant failed

ccl/serverccl.TestResetIndexUsageStatsRPCForTenant [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3658801&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3658801&tab=artifacts#/) on master @ [db229cad354c2776ea35c9dc4f78e6ce66b9437c](https://github.com/cockroachdb/cockroach/commits/db229cad354c2776ea35c9dc4f78e6ce66b9437c):


```
=== RUN   TestResetIndexUsageStatsRPCForTenant
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant089749843
    test_log_scope.go:80: use -show-logs to present logs inline
    tenant_status_test.go:393: 
        	Error Trace:	tenant_status_test.go:393
        	Error:      	Not equal: 
        	            	expected: 0x0
        	            	actual  : 0x1
        	Test:       	TestResetIndexUsageStatsRPCForTenant
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant089749843
--- FAIL: TestResetIndexUsageStatsRPCForTenant (3.12s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/server lindseyjin
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResetIndexUsageStatsRPCForTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/serverccl.TestResetIndexUsageStatsRPCForTenant [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3667786&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3667786&tab=artifacts#/) on master @ [f69253e3d1cfe798821ea1b1c85aeb79e9eece32](https://github.com/cockroachdb/cockroach/commits/f69253e3d1cfe798821ea1b1c85aeb79e9eece32):


```
=== RUN   TestResetIndexUsageStatsRPCForTenant
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant454704883
    test_log_scope.go:80: use -show-logs to present logs inline
    tenant_status_test.go:393: 
        	Error Trace:	tenant_status_test.go:393
        	Error:      	Not equal: 
        	            	expected: 0x0
        	            	actual  : 0x1
        	Test:       	TestResetIndexUsageStatsRPCForTenant
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant454704883
--- FAIL: TestResetIndexUsageStatsRPCForTenant (3.05s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResetIndexUsageStatsRPCForTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/serverccl.TestResetIndexUsageStatsRPCForTenant [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3674762&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3674762&tab=artifacts#/) on master @ [05d36621ae805e644d98ef6b35d453ae16255ccc](https://github.com/cockroachdb/cockroach/commits/05d36621ae805e644d98ef6b35d453ae16255ccc):


```
=== RUN   TestResetIndexUsageStatsRPCForTenant
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant839785845
    test_log_scope.go:80: use -show-logs to present logs inline
    tenant_status_test.go:393: 
        	Error Trace:	tenant_status_test.go:393
        	Error:      	Not equal: 
        	            	expected: 0x0
        	            	actual  : 0x1
        	Test:       	TestResetIndexUsageStatsRPCForTenant
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant839785845
--- FAIL: TestResetIndexUsageStatsRPCForTenant (3.34s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResetIndexUsageStatsRPCForTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/serverccl.TestResetIndexUsageStatsRPCForTenant [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3681125&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3681125&tab=artifacts#/) on master @ [4493953f80f6e79fc461d6a9e6887d81e7ccccae](https://github.com/cockroachdb/cockroach/commits/4493953f80f6e79fc461d6a9e6887d81e7ccccae):


```
=== RUN   TestResetIndexUsageStatsRPCForTenant
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant300562843
    test_log_scope.go:80: use -show-logs to present logs inline
    tenant_status_test.go:413: 
        	Error Trace:	tenant_status_test.go:413
        	Error:      	Not equal: 
        	            	expected: 0x0
        	            	actual  : 0x1
        	Test:       	TestResetIndexUsageStatsRPCForTenant
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant300562843
--- FAIL: TestResetIndexUsageStatsRPCForTenant (3.32s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResetIndexUsageStatsRPCForTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/serverccl.TestResetIndexUsageStatsRPCForTenant [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3690361&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3690361&tab=artifacts#/) on master @ [46b7efa32c57b15f4ae6e0d75845f63360b62161](https://github.com/cockroachdb/cockroach/commits/46b7efa32c57b15f4ae6e0d75845f63360b62161):


```
=== RUN   TestResetIndexUsageStatsRPCForTenant
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant533059952
    test_log_scope.go:80: use -show-logs to present logs inline
    tenant_status_test.go:413: 
        	Error Trace:	tenant_status_test.go:413
        	Error:      	Not equal: 
        	            	expected: 0x0
        	            	actual  : 0x1
        	Test:       	TestResetIndexUsageStatsRPCForTenant
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant533059952
--- FAIL: TestResetIndexUsageStatsRPCForTenant (3.18s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResetIndexUsageStatsRPCForTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/serverccl.TestResetIndexUsageStatsRPCForTenant [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3693032&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3693032&tab=artifacts#/) on master @ [a0849b11188efd1c58dc339d15f75b03e0762459](https://github.com/cockroachdb/cockroach/commits/a0849b11188efd1c58dc339d15f75b03e0762459):


```
=== RUN   TestResetIndexUsageStatsRPCForTenant
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant463324408
    test_log_scope.go:80: use -show-logs to present logs inline
    tenant_status_test.go:413: 
        	Error Trace:	tenant_status_test.go:413
        	Error:      	Not equal: 
        	            	expected: 0x0
        	            	actual  : 0x1
        	Test:       	TestResetIndexUsageStatsRPCForTenant
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestResetIndexUsageStatsRPCForTenant463324408
--- FAIL: TestResetIndexUsageStatsRPCForTenant (3.16s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResetIndexUsageStatsRPCForTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/serverccl/tenant_status_test.go
+++ b/pkg/ccl/serverccl/tenant_status_test.go
@@ -37,6 +37,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
@@ -340,8 +341,9 @@ func TestResetIndexUsageStatsRPCForTenant(t *testing.T) {
 
 	testingCluster := testHelper.testCluster()
 	controlCluster := testHelper.controlCluster()
+	var testingTableID, controlTableID string
 
-	for _, cluster := range []tenantCluster{testingCluster, controlCluster} {
+	for i, cluster := range []tenantCluster{testingCluster, controlCluster} {
 		// Create tables and insert data.
 		cluster.tenantConn(0).Exec(t, `
 CREATE TABLE test (
@@ -366,6 +368,13 @@ VALUES (1, 10, 100), (2, 20, 200), (3, 30, 300)
 		testTableID, err := strconv.Atoi(testTableIDStr)
 		require.NoError(t, err)
 
+		// Set table ID outside of loop.
+		if i == 0 {
+			testingTableID = testTableIDStr
+		} else {
+			controlTableID = testTableIDStr
+		}
+
 		// Wait for the stats to be ingested.
 		require.NoError(t,
 			idxusage.WaitForIndexStatsIngestionForTest(statsIngestionNotifier, map[roachpb.IndexUsageKey]struct{}{
@@ -389,40 +398,62 @@ SELECT
 FROM
   crdb_internal.index_usage_statistics
 WHERE
-  table_id = $1
-`
+  table_id = ` + testTableIDStr
 		// Assert index usage data was inserted.
-		actual := cluster.tenantConn(2).QueryStr(t, query, testTableID)
 		expected := [][]string{
 			{testTableIDStr, "1", "1", "true"},
 			{testTableIDStr, "2", "1", "true"},
 		}
-		require.Equal(t, expected, actual)
+		cluster.tenantConn(2).CheckQueryResults(t, query, expected)
 	}
 
 	// Reset index usage stats.
+	timePreReset := timeutil.Now()
 	status := testingCluster.tenantStatusSrv(1 /* idx */)
 	_, err := status.ResetIndexUsageStats(ctx, &serverpb.ResetIndexUsageStatsRequest{})
 	require.NoError(t, err)
 
+	// Check that last reset time was updated for test cluster.
 	resp, err := status.IndexUsageStatistics(ctx, &serverpb.IndexUsageStatisticsRequest{})
 	require.NoError(t, err)
-
-	// Require index usage metrics to be reset.
-	for _, stats := range resp.Statistics {
-		require.Equal(t, uint64(0), stats.Stats.TotalReadCount)
-		require.Equal(t, time.Time{}, stats.Stats.LastRead)
-	}
+	require.True(t, resp.LastReset.After(timePreReset))
 
 	// Ensure tenant data isolation.
+	// Check that last reset time was not updated for control cluster.
 	status = controlCluster.tenantStatusSrv(1 /* idx */)
 	resp, err = status.IndexUsageStatistics(ctx, &serverpb.IndexUsageStatisticsRequest{})
 	require.NoError(t, err)
+	require.Equal(t, resp.LastReset, time.Time{})
 
-	// Require index usage metrics to not be reset.
-	for _, stats := range resp.Statistics {
-		require.NotEqual(t, uint64(0), stats.Stats.TotalReadCount)
-		require.NotEqual(t, time.Time{}, stats.Stats.LastRead)
+	// Query to fetch index usage stats. We do this instead of sending
+	// an RPC request so that we can filter by table id.
+	query := `
+SELECT
+  table_id,
+  total_reads,
+  last_read
+FROM
+  crdb_internal.index_usage_statistics
+WHERE
+  table_id = $1
+`
+
+	// Check that index usage stats were reset.
+	rows := testingCluster.tenantConn(2).QueryStr(t, query, testingTableID)
+	require.NotNil(t, rows)
+	for _, row := range rows {
+		require.Equal(t, row[1], "0", "expected total reads for table %s to be reset, but got %s",
+			row[0], row[1])
+		require.Equal(t, row[2], "NULL", "expected last read time for table %s to be reset, but got %s",
+			row[0], row[2])
+	}
+
+	// Ensure tenant data isolation.
+	rows = controlCluster.tenantConn(2).QueryStr(t, query, controlTableID)
+	require.NotNil(t, rows)
+	for _, row := range rows {
+		require.NotEqual(t, row[1], "0", "expected total reads for table %s to not be reset, but got %s", row[0], row[1])
+		require.NotEqual(t, row[2], "NULL", "expected last read time for table %s to not be reset, but got %s", row[0], row[2])
 	}
 }
 

--- a/pkg/server/server_sql.go
+++ b/pkg/server/server_sql.go
@@ -762,6 +762,7 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 	)
 
 	distSQLServer.ServerConfig.SQLStatsController = pgServer.SQLServer.GetSQLStatsController()
+	distSQLServer.ServerConfig.IndexUsageStatsController = pgServer.SQLServer.GetIndexUsageStatsController()
 
 	// Now that we have a pgwire.Server (which has a sql.Server), we can close a
 	// circular dependency between the rowexec.Server and sql.Server and set

--- a/pkg/server/tenant_status.go
+++ b/pkg/server/tenant_status.go
@@ -741,6 +741,9 @@ func (t *tenantStatusServer) IndexUsageStatistics(
 		return nil, err
 	}
 
+	// Append last reset time.
+	resp.LastReset = t.sqlServer.pgServer.SQLServer.GetLocalIndexStatistics().GetLastReset()
+
 	return resp, nil
 }
 

--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -461,6 +461,12 @@ func (s *Server) GetSQLStatsController() *persistedsqlstats.Controller {
 	return s.sqlStatsController
 }
 
+// GetIndexUsageStatsController returns the idxusage.Controller for current
+// sql.Server's index usage stats.
+func (s *Server) GetIndexUsageStatsController() *idxusage.Controller {
+	return s.indexUsageStatsController
+}
+
 // GetSQLStatsProvider returns the provider for the sqlstats subsystem.
 func (s *Server) GetSQLStatsProvider() sqlstats.Provider {
 	return s.sqlStats

--- a/pkg/sql/sem/builtins/builtins_test.go
+++ b/pkg/sql/sem/builtins/builtins_test.go
@@ -469,6 +469,35 @@ func TestExtractTimeSpanFromTimeTZ(t *testing.T) {
 	}
 }
 
+// TestResetIndexUsageStatsOnRemoteSQLNode asserts that the built-in for
+// resetting index usage statistics works when it's being set up on a remote
+// node via DistSQL.
+func TestResetIndexUsageStatsOnRemoteSQLNode(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+
+	testCluster := serverutils.StartNewTestCluster(t, 3 /* numNodes */, base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{},
+	})
+	defer testCluster.Stopper().Stop(ctx)
+	testConn := testCluster.ServerConn(2 /* idx */)
+	sqlDB := sqlutils.MakeSQLRunner(testConn)
+
+	query := `
+CREATE TABLE t (k INT PRIMARY KEY);
+INSERT INTO t SELECT generate_series(1, 30);
+
+ALTER TABLE t SPLIT AT VALUES (10), (20);
+ALTER TABLE t EXPERIMENTAL_RELOCATE LEASE SELECT 1, 1;
+ALTER TABLE t EXPERIMENTAL_RELOCATE LEASE SELECT 2, 15;
+ALTER TABLE t EXPERIMENTAL_RELOCATE LEASE SELECT 3, 25;
+
+SELECT count(*) FROM t WHERE crdb_internal.reset_index_usage_stats();
+`
+
+	sqlDB.Exec(t, query)
+}
+
 func TestExtractTimeSpanFromInterval(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 

